### PR TITLE
Reconnect broken LDAP connections on ldb_wait fails

### PR DESCRIPTION
**Pull request by exekias**

### CODE REVIEW

**Comment by exekias** [comment]
jenkins retest this please

File: mapiproxy/util/samdb.c [hunk 1]
```diff
@@ -211,3 +211,51 @@ int safe_ldb_search(struct ldb_context **ldb_ptr, TALLOC_CTX *mem_ctx,
...
+				continue;
+			}
+
+			ret = ldb_wait(req->handle, type);
+			if (ret != LDB_ERR_OPERATIONS_ERROR) {
+				/* reconnect worked, replace ldb_context with the new one */
+				talloc_free(*ldb_ptr);
+				*ldb_ptr = new_ldb;
+				break;
+			}
```
Comment (blaxter) [inline mapiproxy/util/samdb.c]: no need `\n`

**Comment by blaxter** [comment]
it should be ok :+1:

File: mapiproxy/util/samdb.c [hunk 2]
```diff
@@ -211,3 +211,51 @@ int safe_ldb_search(struct ldb_context **ldb_ptr, TALLOC_CTX *mem_ctx,
...
+
+	ret = ldb_wait(req->handle, type);
+	if (ret != LDB_ERR_OPERATIONS_ERROR) {
+		return ret;
+	}
+
+	/* Something failed, try reconnecting */
+	do {
+		OC_DEBUG(3, "sam db connection lost, reconnecting...");
+		new_ldb = samdb_init(talloc_parent(*ldb_ptr));
```
Comment (sixstone-qq) [inline mapiproxy/util/samdb.c]: It will crash if `ldb_ptr` is NULL, do you think sanity checks are required at the beginning of the function?

Comment (exekias) [inline mapiproxy/util/samdb.c]: :+1: I added it at line 226

**Comment by sixstone-qq** [comment]
LGTM

**Pull request merged**